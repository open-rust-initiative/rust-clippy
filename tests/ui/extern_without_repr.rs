#![allow(unused)]
#![allow(improper_ctypes_definitions)]
#![allow(improper_ctypes)]
#![warn(clippy::extern_without_repr)]
#![feature(rustc_private)]
#![feature(core_intrinsics)]
extern crate libc;

#[repr(C)]
struct Foo1 {
    a: i32,
    b: i64,
    c: u128,
}

#[repr(packed)]
#[derive(Debug, Clone, Copy)]
struct Foo2 {
    a: libc::c_char,
    b: libc::c_int,
    c: libc::c_longlong,
}

struct Foo3 {
    a: libc::c_char,
    b: libc::c_int,
    c: libc::c_longlong,
}

extern "C" fn c_abi_fn1(arg_one: u32, arg_two: usize) {}
extern "C" fn c_abi_fn2(arg_one: u32, arg_two: Foo1) {}
extern "C" fn c_abi_fn3(arg_one: u32, arg_two: *const Foo2) {}
extern "C" fn c_abi_fn4(arg_one: u32, arg_two: *const Foo3) {}

extern "C" {
    fn c_abi_in_block1(arg_one: u32, arg_two: usize);
    fn c_abi_in_block2(arg_one: u32, arg_two: Foo1);
    fn c_abi_in_block3(arg_one: u32, arg_two: Foo2);
    fn c_abi_in_block4(arg_one: u32, arg_two: Foo3);
}

fn main() {
    // test code goes here
}
