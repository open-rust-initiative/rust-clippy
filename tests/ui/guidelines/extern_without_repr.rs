#![allow(unused)]
#![allow(improper_ctypes, improper_ctypes_definitions)]
#![warn(clippy::extern_without_repr)]
#![feature(rustc_private)]
#![feature(repr_simd, simd_ffi)]
extern crate libc;

use std::ffi::c_int;

#[repr(C)]
struct ReprC {
    a: i32,
    b: i64,
}

#[repr(packed)]
#[derive(Debug, Clone, Copy)]
struct Packed {
    a: libc::c_char,
    b: libc::c_int,
    c: libc::c_longlong,
}

#[repr(align(2))]
struct Aligned {
    inner: c_int,
}

#[repr(simd)]
struct ReprSimd {
    inner: c_int,
}

#[repr(transparent)]
struct Transparent {
    inner: c_int,
}

struct NakedStruct {
    a: libc::c_char,
}

pub enum NakedEnum {
    A,
    B,
}

extern "C" fn c_abi_fn1(arg: c_int) {}
extern "C" fn c_abi_fn2(arg_one: c_int, arg_two: ReprC) {}
extern "C" fn c_abi_fn3(arg_one: c_int, arg_two: *const Packed) {}
extern "C" fn c_abi_fn4(arg: Aligned) {}
extern "C" fn c_abi_fn5(arg: ReprSimd) {}
extern "C" fn c_abi_fn6(arg: Transparent) {}
extern "C" fn bad1(arg_one: c_int, arg_two: *const NakedStruct) {}
extern "C" fn bad2(arg: NakedStruct) {}
extern "C" fn bad3(arg: NakedEnum) {}
extern "C" fn bad4(arg: NakedEnum, arg2: NakedStruct) {}

extern "C" {
    fn c_abi_in_block1(arg_one: c_int);
    fn c_abi_in_block2(arg_one: c_int, arg_two: ReprC);
    fn c_abi_in_block3(arg_one: c_int, arg_two: Packed);
    fn c_abi_in_block4(arg: Aligned);
    fn c_abi_in_block5(arg: ReprSimd);
    fn bad_in_block1(arg_one: c_int, arg_two: NakedStruct);
    fn bad_in_block2(arg: *mut NakedEnum);
    fn bad_in_block3(arg: *mut NakedEnum, arg2: NakedStruct, arg3: *const ReprSimd);
}

fn main() {}
