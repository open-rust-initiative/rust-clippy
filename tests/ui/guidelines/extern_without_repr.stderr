error: Should use repr to specifing data layout when struct is used in FFI
  --> $DIR/extern_without_repr.rs:24:1
   |
LL | struct Foo3 {
   | ^^^^^^^^^^^
   |
   = note: `-D clippy::extern-without-repr` implied by `-D warnings`

error: aborting due to previous error

